### ISSUE TITLE
Chrome returning 400, Firefox working fine

### ISSUE THREAD

**Issue by mselevan** [web-push-libs/pywebpush#47]
Using chrome version 57. Get 400s with chrome and not on Firefox. Have tried just about everything under the sun. I know there have complaints about this before. Any solutions to the problems before?

**Comment by jrconlin** [comment]
400 is an interesting error message to get back, and not a lot to work off of. 

Can you provide the full message response? There may be additional information in it that could help.

**Comment by mselevan** [comment]
So I was able to resolve the issue with chrome. Quite the nuance. The aud field in the vapid claims cannot have a '/' at the end of the Firebase (fcm) URL. Firefox handles it fine but fcm seemed to not like that little detail. As soon as I removed that '/' from the end of the URL everything worked fine. Thought I would share. Maybe something that can be chopped off by the library in future releases. 

Thanks again for all your help!

**Comment by jrconlin** [comment]
Interesting.

FWIW, URLs are supposed to be very specific, and (pardon if I get a bit nerdy here) 

"http://example.com/foo" refers to the "foo" file under the "root" and "http://example.com/foo/" refers to the index file under "/foo". Of course, things get really, really fuzzy with "http://example.com" since that doesn't refer to any file specifically, but defaults to be the index in the root. 

The "aud" refers to the target "audience", it's a bit of a reversal of things and means the intended recipient of the VAPID certificate (rather than the URL of what to pull off of the server). Google (being google) is super pedantic about things and wants to make absolutely sure that you address the "aud" to the exact recipient. Mozilla, frankly, doesn't really care. 

VAPID is about the owner identifying themselves. The most important thing in there is the "sub", since that's how we could get in touch with you if we needed to. We check that "you" are "you" by doing a signature check on the key, and doing a bit of equal magic on the URL. Still, the underlying JWT requirement (which is one of the dependencies for VAPID) requires that:
1) The "aud" match the schema and host of the recipient (so, yeah, not the slash, since that's the index file on the host)
2) The "exp" for that VAPID transaction be no more than 24 hours. (that's just the individual transaction, not the VAPID key you generated. That can live as long as you like.)
3) The "sub" be a valid URL construct. (that's why you use "mailto:bob@example.com" instead of "bob@example.com". I don't know why you'd ever do it, but "http://example.com/contact/bob" would also be valid for the spec, but probably not that useful for the poor ops person trying to contact Bob.)

In short:
So, yeah, having the library "fix" any value could actually make things worse if some provider did something different or changed the "aud" requirement. I think it's absolutely a good idea to check and recheck anything you're sending.

**Comment by doandu** [comment]
@mselevan thank you very much, this works for me. I struggled with this for days. How could you find this?

**Comment by KntNico** [comment]
Hi, I've the same issue but unfortunately I'didn't resolved. 


`
o = urlparse(subscription['endpoint'])
aud = o.scheme + '://' + o.netloc

vapid_headers = self.pV.sign({"aud": aud,
                                    "sub": "mailto:myemail@gmail.com",                                    
                                    "exp": int(time.time()) + 720                                                                 
                                    })          
            
            x = WebPusher(subscription_info = {"endpoint":subscription["endpoint"],
                                         "keys": {"auth":subscription["keys"]["auth"],
                                                  "p256dh":subscription["keys"]["p256dh"]
                                                  }
                                         }
                          ).send(payload['subject'] + ' ' +payload['body'],
                                 vapid_headers,
                                 ttl= 60 * 60 * 2                                                                  
                                 )`

still returns 400s unauthorized on Chrome, works fine with Mozilla
py_vapid 1.2.1 
pywebpush 0.8.0

Could it be caused by the "gcm_sender_id" in the manifest? 
Thanks

**Comment by jrconlin** [comment]
Hmm. I'm able to reproduce the bug on my side (which is helpful) but not really sure what the cause might be. Give me a while to figure that out.

Thanks!

**Comment by jrconlin** [comment]
Ok,, so looks like I've isolated the problem. Chrome appears to be sorting the VAPID claims before checking the signatures. Unfortunately, this is a bug according to the [JWS spec](https://tools.ietf.org/html/rfc7515#section-5.1). I've contacted the folks over at Google about it, but for now, VAPID 1.2.3 will sort the claims. Since it's really bad form to push libraries out late on Fridays, I'll promote things Monday morning PST

**Comment by KntNico** [comment]
Ok thank you! I'll wait for the update and I will test it on my side once ready.

**Comment by jrconlin** [comment]
Ok, so, here's a fun fact I just recently learned. 

If you're using the Chrome developer console and calling something like 
`JSON.stringify(subscription.toJSON())` to get a JSON serialization of the WebPush subscription info, Chrome may helpfully inject a `…` into the endpoint URL inside of the JSON string. Unfortunately this will make the endpoint invalid and cause GCM/FCM to return a 400 error. 

If you're testing things by exporting the endpoint, be certain to check that you're actually getting the entire, unaltered endpoint. Likewise each of the "keys" should be proper base64 strings (although padding may or may not be present)

I've verified that the current version of pywebpush works with Chrome Canary and may push a minor update soon to verify that the endpoint URL does not contain invalid characters.

**Comment by Flimm** [comment]
Hmm, the `…` ellipsis should only be visible to the end-user when using the Chrome developer tools, it's not actually changing the endpoint string in the object, right? That would be truly evil.

**Comment by kostasgeo** [comment]
Hello, 
The ... would of course be a problem, but I can verify the 400 error on Chrome with a correct and subscribed endpoint using this library. I have successfully sent a notification to the exact same endpoint on Chrome using another library (https://github.com/web-push-libs/web-push). As seen in similar posts, the library is successful for Firefox.

**Comment by jrconlin** [comment]
No, it's not changing the endpoint within the object, only the output string, thus why it's being "helpful". I was using Matt Gaunt's demo at "https://gauntface.github.io/simple-push-demo/", stopping it within the debugger, and copying the result of a `JSON.stringify(subscription.toJSON())`. 
pywebpush can use the contents of that string as the --info parameter. I had simply copied the console output en block and written it to a file. Sometimes it would work in chrome, sometimes it wouldn't, but a similar technique would always work in firefox.

A bit of help from my counterpart pointed out a '…' in the endpoint URL. I worked backwards and found the source. Since it's a single character in a long complex string, it might be easy to miss.

**Comment by jrconlin** [comment]
@kostasgeo, I would be very interested in seeing the data that you're using since I'm unable to reproduce the problem. 

In my case, I am using
claims:
`{"aud":"https://fcm.googleapis.com", "sub": "mailto:invalid@jrconlin.com"}`
info:
`{"endpoint":"https://fcm.googleapis.com/fcm/send/ffMTaCDnb6k:APA91bHtwhMXGDNFry9PYTJjb6VyoFO1rk4FXYNh37_IYG1YZtsg8_o6obDncDzNriXKl9rlZ3EFhl19rBStTMjDb5361LbD6ryLaOSvSrr_QP30QffKgBfyNZ4WNbto_KRMFPADC6hn","keys":{"p256dh":"BCphCAfIxyT-4fSAwjyUNllUAh9bod20V13qCHi9GKAcmGhTj9O5jBUEfkUlKPsastlrBoOayuY8E7UfK2WF1o4=","auth":"IiBPSp_Ku-A_1905kSeq4Q=="}}`
and the key data from https://gauntface.github.io/simple-push-demo/

When I run 
```
pywebpush  --info chrome_info.json --claims chrome_claims.json --key `cat chrome_key.raw` --data test_data.txt
```
I get a 201 and a chrome notification. I've disabled notifications to that endpoint, so I don't know if the URL will still work, but you can use it to at least double check the endpoint you're sending to. I believe that a 400 error may be a default return for any number of errors.

Can you include the subscription info, claims, and key data and send it as an email to jr@mozilla.com? I'm going to have limited availability the next few days, but will try to figure out what may be the error.

**Comment by kostasgeo** [comment]
@jrconlin, I have sent you the data requested.

**Comment by KntNico** [comment]
@jrconlin I edited the py_vapid library in my development environment with the claims sort fix and chrome now returns 200 and works fine. When py_vapid 1.2.3 will be released on pypi?  

Thank you!

**Comment by jrconlin** [comment]
Hmm. I hadn't pushed that out mostly because the argument order shouldn't have made any difference. I've pushed out py_vapid 1.2.3  just now.

**Comment by kostasgeo** [comment]
Everything seems fine with py_vapid 1.2.3. Thanks a lot !

**Comment by jrconlin** [comment]
RESOLVED
